lib/fluent/plugin/kubernetes_metadata_cache_strategy.rb

```diff
@@ -0,0 +1,91 @@
+#
+# Fluentd Kubernetes Metadata Filter Plugin - Enrich Fluentd events with
+# Kubernetes metadata
+#
+# Copyright 2017 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+module KubernetesMetadata
+  module CacheStrategy
+
+    def get_pod_metadata(key, namespace_name, pod_name, record_create_time)
+      metadata = {}
+      ids = @id_cache[key]
+      if !ids.nil?
+        # FAST PATH
+        # Cache hit, fetch metadata from the cache
+        metadata = @cache.fetch(ids[:pod_id]) do
+          @stats.bump(:pod_cache_miss)
+          m = fetch_pod_metadata(namespace_name, pod_name)
+          (m.nil? || m.empty?) ? {'pod_id'=>ids[:pod_id]} : m
+        end
+        metadata.merge!(@namespace_cache.fetch(ids[:namespace_id]) do
+          @stats.bump(:namespace_cache_miss)
+          m = fetch_namespace_metadata(namespace_name)
+          (m.nil? || m.empty?) ?  {'namespace_id'=>ids[:namespace_id]} : m
+        end)
+      else
+        # SLOW PATH
+        @stats.bump(:id_cache_miss)
+        pod_metadata = fetch_pod_metadata(namespace_name, pod_name)
+        namespace_metadata = fetch_namespace_metadata(namespace_name)
+        ids = { :pod_id=> pod_metadata['pod_id'], :namespace_id => namespace_metadata['namespace_id'] }
+        if !ids[:pod_id].nil? && !ids[:namespace_id].nil?
+          # pod found and namespace found
+          metadata = pod_metadata
+          metadata.merge!(namespace_metadata)
+        else
+          if ids[:pod_id].nil? && !ids[:namespace_id].nil?
+            # pod not found, but namespace found
+            @stats.bump(:id_cache_pod_not_found_namespace)
+            ns_time = Time.parse(namespace_metadata['creation_timestamp'])
+            if ns_time <= record_create_time
```

Comment:
Is there a test for this case?

Reply:
This should be the testcase that satisfies this path: https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter/pull/103/files#diff-6c11ec708e3e636a7c0eead24e437db5R128